[MERGE] add new module product_lifecycle that manage the data model of the replacement of discontinued products

zaoral opened the pull request:
Task [vx#3453](https://www.vauxoo.com/#id=3453&view_type=form&model=project.task)
Task [vx#3649](https://www.vauxoo.com/#id=3649&view_type=form&model=project.task)

 HU [314](https://www.vauxoo.com/?debug=#id=314&view_type=form&model=user.story&action=726) Sprint 12

Runbot: Manual Pre-Build  http://runbot.vauxoo.com/runbot/team/yoytec-24#102

zaoral committed:
[ADD] add new module basic structure.

zaoral committed:
[ADD] add new field replacement_product_id to the product.template model and
view.

zaoral committed:
[ADD] The Status of the product now is manage in the header of the product
form view as a clickable statusbar widget and with default value `In
Development`.

zaoral committed:
[ADD] some search filters were added to search and group by products lifecycle
status.

zaoral committed:
[ADD] add the documentation of the module in description index.html file and
update the README file.

zaoral commented:
waiting for testing tools results

zaoral committed:
[ADD] Add a rule to restrict the edit a product. Only the
product manager can edit the product and the product variant.

zaoral committed:
[ADD] new wizard that extract the replacement products of a discontinued
product. Show then and let the user to select one.

zaoral committed:
[ADD] do not let to create a a purchase order with a discontinued product.
will reaise a warning indicating that the product is discontinued.

zaoral committed:
[ADD] add new logic in the purchase order lines. When the user add a
discontinued product will indicate a warning with this information and will
require to the user to select a replacement product.

zaoral committed:
[ADD] add check the purchase order line to look for discontinued products in
the lines, if is the case will raise a warning indicating that the purchase
order can not have purchase order lines with discontinued products.

zaoral committed:
[ADD] the functionality was moved from the product.template model to the
product.product model.

zaoral committed:
[ADD] add label to identificate that the replacement product are for purchase
operations.

zaoral committed:
[ADD] add demo data for product states and its replacement products.

zaoral committed:
[ADD] only show as a replacement the products that are not obsolete and active

zaoral committed:
Merge branch '8.0' of github.com:Vauxoo/addons-vauxoo into 8.0-vx#3453-kty

zaoral committed:
[ADD] add sequence to the purchase order and show the line sequence in the
message of discontinued product.

zaoral committed:
[pylint][flake8] Multiple Fixes:

[pylint]
- W0622 redefinition of Warning import from openerp.

[flake8]
- E126 continuation line over-indented for hanging indent
- E231 missing whitespace after ','

zaoral committed:
[ADD] hide the status state in the product template. this field is not used
anymore. the new state2 state defined in product.product is what defines the
status of the variant.

zaoral committed:
[ADD] show the status of every variant in the product.template form at the
Inventory > Status section. This is only show if watching from the product
template form but not form the product variant form.

zaoral committed:
i[ADD] Replacement Product in the purchase order line was deleted to use a
new field Discontinued Product. This new field holds the obsolete
product and when is fill will add a domain into the Product field with the
good replacements products (active and not obsolete products marked as a
replacements of the current discontinued product). In other case, if the user
try to fill the Product field with a discontinued product will show a warning
indicating that can not use a discontinued product in the purchase order and
will move this value to the Discontinued Product field, cleaning the Product
field so you can select a replacement product.

In a technical matters, some methods were modularizate to optimize code, the
methods docstring were updated to be more detail and the old comments in code
were deleted.

zaoral committed:
[ADD] show the display name of the variant instead of the code.

zaoral committed:
[ADD] update description of the module.

zaoral committed:
[ADD] change the replacement wizard product to be call from the purchase order
and extract the information of the discontinued products in the order lines to
be able to select the replacements configured in the product. The purchase
orders will be updated will move the product to the discontinued product field
and will add the new replacement selected in the wizard to the product field.

zaoral committed:
[ADD] new module product_lifecycle translations. Some fixes to the code was
made.

zaoral committed:
[ADD] update the module help indicating how it works the Replace Discontinued
Product Wizard.

zaoral commented:
@nhomar please review this PR

travis is red, but for external reasons to this PR

#### Note: This PR must be merge after Vauxoo/addons-vauxoo#291 is merge

zaoral commented:
I new issue odoo/odoo#6994 was made to discuss change of the product.template state field to the product.product model.

nhomar commented on product_lifecycle/__openerp__.py:
```diff
@@ -0,0 +1,52 @@
+    "depends": [
+        "product",
+        "purchase",
+        "purchase_order_line_sequence",
```
No necesitas las dos dependencias de arriba. purchase_order_line sequence depende de purchase y a su vez este de product.

Por otra parte no entiendo por que purchase_order_line_sequence es necesaria para éste feature, me comentas por que?

Saludos.

zaoral replied:
@nhomar 
- I will remove the unnecessary dependencies.
- This functionality we use it on the product_lifecycle module that need for the identificate the purchase order lines to replace and to indicate to the user what lines have discontinued products.

nhomar commented:
@zaoral Please:
1. Add unit test following all the product lifecycle to ensure always this module is tested mixed with all others which manipulate products.
2. Can you make a video with this feature working please?

Regards.

zaoral commented:
@nhomar ok. working on the unit test and the video. Please approve this guideline for the video https://drive.google.com/a/vauxoo.com/file/d/0B1UUUA6bRx8-MzRCc2ZBN054Tkk/view

zaoral committed:
[ADD] the new state2 field in the product.product module now have the
copy=False attribute to make sure than when make a copy do not copy the field
just add the default.

zaoral committed:
[ADD] Improves over the Replacement Wizard:
- [ADD] Add constraint to do not let to save a replacement line if the
  discontinued product is not a obsolete product or if the replacement
  product is a obsolete product.
- [FIX] resolve pylint error about importing openerp.exceptions.Warning
- [CC] reorganizate the imports taking into accounv v8 style.

zaoral committed:
[ADD] update the demo data to construct better unit test cases.

zaoral committed:
[ADD] add unti test to the product_lifecycle module.

zaoral committed:
Merge pull request #6 from vauxoo-dev/8.0-vx#3649-kty

[ADD] add unit test to the product_lifecycle module.

zaoral commented:
A merge with the new unit test was made to this branch.

Waiting for runbot result in http://runbot.vauxoo.com/runbot/team/yoytec-24#48pull_266
At this moment travis result is red it have importing errors external to this PR

zaoral committed:
[ADD] TODO Comment testing related to demo data because changes made have been
commited to the database. This generate a unit test error when running test a
second time. This was report to odoo. after is fixes uncomment the section of
code.

zaoral committed:
Merge branch '8.0' of github.com:Vauxoo/addons-vauxoo into 8.0-vx#3453-kty

zaoral committed:
[pylint] Fix importing warnings.

zaoral commented:
Waiting for combined runbot result http://runbot.vauxoo.com/runbot/team/yoytec-24#102

zaoral committed:
Merge branch '8.0' of github.com:Vauxoo/addons-vauxoo into 8.0-vx#3453-kty

zaoral committed:
[FIX] resolve problem with the unit test. The test was not doing a rollback to
the end of the test so when trying to run the test after the first time the
database was corrupted and the test return errors. Change the schema of the
test from a SingleTransactionCase to a TransactionCase. Now every test method
is separate. Some changes to modularizate the test were made to fullfile the
invidivuality per test method. Also the not functional commented test were
uncommented and are working propertly.

zaoral committed:
Merge pull request #8 from vauxoo-dev/8.0-vx#3649-kty

[FIX] resolve problem with the unit test. The test was not doing a rollback to the end of the test so when trying to run the test after the first time the database was corrupted and the test return errors. Change the schema of the test from a SingleTransactionCase to a TransactionCase. Now every test method is separate. Some changes to modularizate the test were made to fullfile the invidivuality per test method. Also the not functional commented test were uncommented and are working propertly.

zaoral committed:
[FIX] when no order lines muts return a empty list not a boolean value. this
was generating and error in the test of yoytec_customer_rma_workflow module.

zaoral committed:
Merge branch '8.0-vx#3453-kty' of github.com:vauxoo-dev/addons-vauxoo into 8.0-vx#3453-kty

zaoral commented:
@nhomar PR ready. Runbot green http://runbot.vauxoo.com/runbot/team/yoytec-24#102 for this PR and for https://github.com/Vauxoo/yoytec/pull/124. Please review this two PRs. First this one and next the https://github.com/Vauxoo/yoytec/pull/124
